### ISSUE TITLE
Group Thin-Edge-Json and MQTT definitions in an `tedge_mqtt_api` crate

### ISSUE THREAD

**Issue by didier-wenzek** [thin-edge/thin-edge.io#1410]
This refactoring task is part of https://github.com/thin-edge/thin-edge.io/issues/1401.

## Motivation

With the current code base, the MQTT protocol of thin-edge is scattered over several crates,
mainly `thin_edge_json` and `agent_interface` but also `c8y_configuration_plugin` that defines `tedge/+/commands/res/config_snapshot` and `tedge/+/commands/res/config_update`,
or even `tedge_agent` if we consider the HTTP api of the agent as being part of the public API of thin-edge.

This HTTP API is also undocumented.
We are missing an exhaustive list of topics and messages that define the interface with thin-edge.

## Proposal

The proposal is to have a single crate, named `tedge_mqtt_api` to gather all the definitions related to the MQTT+HTTP interface of thin-edge.

Expectations:

* A `tedge_mqtt_api` crate that groups all the definitions currently scattered between `thin_edge_json` and `agent_interface`
  * the MQTT API (topic names, message payloads),
  * the HTTP API (uri, verbs, content),
* Updated `mapper`, `agent` and plugins using the `tedge_mqtt_api`.
* A reference document listing
  * All the `tedge` topics with their purpose and the expected payloads.

Out of scope:

* Any changes that is visible by a user.
* Improvements of the MQTT API and topic scheme.
  * There are many inconsistent choices for the topic names (mostly related to the child devices).
     The point here is not to fix those but to gather in one place all these definitions
     and to make the differences easier to spot and fix.
* The SW plugin API and the `tedge` cli interface.
   * There are requests to trigger `tedge config` command over MQTT or to have an MQTT api for SW.
   * But this will be done only once the current state properly defined and documented.

**Comment by albinsuresh** [comment]
> This HTTP API is also undocumented.

It is documented now.

> The proposal is to have a single crate, named `tegde_mqtt_api` to gather all the definitions related to the MQTT+HTTP interface of thin-edge.

If we're combining both the MQTT and HTTP APIs into a single crate, then `tedge_api` might be a better name for it. Else have separate `tegde_mqtt_api` and `tegde_http_api` crates.

**Comment by didier-wenzek** [comment]
> It is [documented](https://github.com/thin-edge/thin-edge.io/blob/main/docs/src/references/tedge-file-transfer-service.md) now. But, not in an OpenAPI format. Is that what you were expecting?

No, I don't think an OpenApi is appropriate here. The current doc is clear enough.

> If we're combining both the MQTT and HTTP APIs into a single crate, then tedge_api might be a better name for it.

This might be confused with the API for Rust internal components/actors. That said we can go with `tedge_api` for the MQTT+HTTP and `tedge_actor` for rust components.

**Comment by rina23q** [comment]
> That said we can go with tedge_api for the MQTT+HTTP

I also agree with this. 

All other points look very clear to what we do.

**Comment by reubenmiller** [comment]
fyi: I just fixed some typos of `tegde` => `tedge` in various comments (to improve the searchability)
